src/service/search/datafusion/to_arr_string_udf.rs

```
@@ -0,0 +1,127 @@
+
+use std::sync::Arc;
+
+use arrow::array::StringArray;
+use arrow_schema::Field;
+use config::utils::json;
+use datafusion::{
+    arrow::{array::ArrayRef, datatypes::DataType},
+    common::cast::{as_generic_list_array, as_generic_string_array},
+    error::DataFusionError,
+    logical_expr::{ScalarUDF, Volatility},
+    prelude::create_udf,
+    sql::sqlparser::parser::ParserError,
+};
+use datafusion_expr::ColumnarValue;
+use once_cell::sync::Lazy;
+
+/// The name of the to_array_string UDF given to DataFusion.
+pub const TO_ARR_STRING_UDF_NAME: &str = "to_array_string";
+
+/// Implementation of to_array_string
+pub(crate) static TO_ARR_STRING: Lazy<ScalarUDF> = Lazy::new(|| {
+    create_udf(
+        TO_ARR_STRING_UDF_NAME,
+        // takes one argument: the field
+        vec![DataType::List(Arc::new(Field::new(
+            "item",
+            DataType::Utf8,
+            true,
+        )))],
+        Arc::new(DataType::Utf8),
+        Volatility::Immutable,
+        Arc::new(to_arr_string_impl),
+    )
+});
+
+/// to_array_string function for datafusion
+/// Converts a datafusion array into a json array string (all the array elements are also
+/// stringified)
+pub fn to_arr_string_impl(args: &[ColumnarValue]) -> datafusion::error::Result<ColumnarValue> {
+    let args = ColumnarValue::values_to_arrays(args)?;
+    log::debug!("Inside cast_to_arr_impl");
+    if args.len() != 1 {
+        return Err(DataFusionError::SQL(
+            ParserError::ParserError("UDF params should be: to_array_string(array)".to_string()),
+            None,
+        ));
+    }
+
+    let list_array = as_generic_list_array::<i32>(&args[0])?;
+
+    let array = list_array
+        .iter()
+        .map(|array| {
+            if let Some(string_array) = array {
+                let string_array = as_generic_string_array::<i32>(&string_array)
+                    .expect("failed to cast to string array from list array");
```

Whats the impact of this expect() call here? 

Can we return the error rather, given you are already returning `datafusion::error::Result<ColumnarValue>` therefore this might make more sense `as_generic_string_array::<i32>(&string_array)?`

---

The `.map` function actually returns an `Option<>`, so using `?` will not work here as `as_generic_string_array()` returns a `Result`.